src/interop/js-interop/js-types.md

```diff
@@ -0,0 +1,212 @@
+
+## Type hierarchy
+
+JS types form a natural type hierarchy:
+
+- Top type: `JSAny`, which is any non-nullish JS value
+  - Primitives: `JSNumber`, `JSBoolean`, `JSString`
+  - `JSSymbol`
+  - `JSBigInt`
+  - `JSObject`, which is any JS object
+    - `JSFunction`
+      - `JSExportedDartFunction`, which represents a Dart callback that was
+      converted to a JS function
+    - `JSArray`
+    - `JSPromise`
+    - `JSDataView`
+    - `JSTypedArray`
+      - JS typed arrays e.g. `JSUint8Array`
+    - `JSBoxedDartObject`, which allows users to box and pass Dart values
+      opaquely within the same Dart runtime
+
+You can find the definition of each type in the [`dart:js_interop` API docs].
+
+{% comment %}
+TODO (srujzs): Refer to this in the "interop type" section in syntax.
+TODO (srujzs): Should we add a tree diagram instead for JS types?
+{% endcomment %}
+
+## Conversions
+
+Since there are two separate domains, you will likely want to *convert* values
+between one domain to the other. For example, you may want to convert a Dart
+`String` into a JS string, which is represented by the JS type `JSString`. To do
+this, Dart supplies a number of `extension` members on various Dart types that
+will do the conversion for you, usually labeled as or prefixed with `toJS`.
+Similarly, you can convert JS values to Dart types. Dart supplies a number of
+members on various JS types to enable this conversion, usually labeled as or
+prefixed with `toDart`. They are often marked `external`, and their
+implementation is usually platform-dependent. Note that not all JS types have a
+conversion, and not all Dart types have a conversion.
+
+In general, the conversion table looks like the following:
+
+| JS type                             | Dart type                                |
+| ----------------------------------- | ---------------------------------------- |
+| `JSNumber`, `JSBoolean`, `JSString` | `num`, `int`, `double`, `bool`, `String` |
+| `JSExportedFunction`                | `Function`                               |
+| `JSArray<T extends JSAny?>`         | `List<T extends JSAny?>`                 |
+| `JSPromise<T extends JSAny?>`       | `Future<T extends JSAny?>`               |
+| Typed arrays e.g. `JSUint8Array`    | `dart:typed_data`                        |
+| `JSBoxedDartObject`                 | Opaque Dart value                        |
+
+**Important**: Conversions may have different costs depending on the compiler.
+Prefer to only convert when you need to. Furthermore, conversions may or may not
+produce a new value. This doesn’t matter for immutable values like numbers, but
+does matter for types like `List`s. A conversion to a `JSArray` may produce a
+new value by copying or may not, so do not rely on modifications to the
+`JSArray` to affect the `List`. Typed array conversions have a similar
+limitation. Look at the specific conversion function for more details.
+
+## Requirements on `external` declarations and `Function.toJS`
```

Comment:
Is this related to the previous section, like kind of a subtopic? If son
```suggestion
### Requirements on `external` declarations and `Function.toJS`
```

Reply:
I think it's a separate topic. Conversions are related, but not all of the information in this sections has to do with conversions.